[Manual Taxes M2] Populate UI with fetched tax rates

toupper opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #10612 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
With this PR we implement the UI population with the fetch tax rates. We also added pagination to the view and the logic to show the bottom notice when all pages are loaded. Furthermore, we tweaked the UI to add some missing parts and extract some views to their own structs.  Apart from that, we fixed some elements in Yosemite and Networking to ensure the right functioning of this feature.
Please note that the empty state is not implemented in this PR to avoid making it even bigger. It will be part of one of the next PRs once the design is finalized.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

With a store with tax rates (please let me know if you want to use one with many tax rates added)

1. Go to Orders
2. Tap on + to create a new order
3. Tap on Set New Tax Rate
4. Make sure that pagination and pull to refresh work fine. Please also check that the _Edit Tax Rates in Admin_ notice is shown at the bottom

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

### Many Pages

https://github.com/woocommerce/woocommerce-ios/assets/1864060/6fb83be4-8744-4a7e-a8cc-8fd09605d8d4

### One page

https://github.com/woocommerce/woocommerce-ios/assets/1864060/8a160c7b-6305-492c-94db-80cfc043918c

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

toupper committed:
Remove tax rates before loading the first page.

toupper committed:
Improve code style

toupper committed:
Add siteID parameter to loadTaxRates

toupper committed:
Add StorageTaxRate typealias

toupper committed:
Add delete tax rates function

toupper committed:
Populate UI with tax rates.

toupper committed:
As names can be duplicated for tax rates, use id as unique identifier.

toupper committed:
Add page parameters

toupper committed:
Only delete tax rates when requesting page one

toupper committed:
Improve next page load

toupper committed:
Attach admin bottom notice to the end of the tax rates list.

toupper committed:
Format percentage

toupper committed:
Tweak design

toupper committed:
Merge branch 'trunk' into issue/10612-populate-ui-tax-rates

# Conflicts:
#	Yosemite/Yosemite/Stores/TaxStore.swift

toupper committed:
Remove not needed comments

toupper committed:
Add tests for NewTaxRateSelectorViewModel

toupper committed:
Decrease line length

iamgabrielma commented:
I haven't started with the review yet, but there are some missing references to `self` in the `sync()` method that makes the build to not compile:
``` diff
- if taxRateViewModels.isEmpty {
+ if self.taxRateViewModels.isEmpty {
...
- } else if results.isEmpty {
+ } else if self.results.isEmpty {
... 
```

toupper committed:
Use self in closure to make CI build pass

joshheald reviewed (commented):
> Go to Orders
> Tap on + to create a new order
> Tap on Set New Tax Rate
> Make sure that pagination and pull to refresh work fine. Please also check that the Edit Tax Rates in Admin notice is shown at the bottom


## Pagination

- Pagination works correctly, I tried with three pages by reducing the per_page to 10. It has to be a long enough list to scroll for pagination to trigger at all though – if I used 5, I couldn't get the second page to load. Since the default is 25, I _think_ that will be OK, and I tried it on a portrait 12.9" iPad simulator, but if there's a way to fix it quickly, it might be nice for futureproofing (e.g. on a Vision Pro, how long will the list be? 😉)
- However it looks like there should be some padding between the bottom row and the page loading spinner.


## Pull to refresh

- This isn't available on lists of tax rates which are too short to scroll. 
- It only works on scrollable lists if you first scroll down, then back to the top
- It doesn't have the usual Pull-to-Refresh "feel", like you're pulling past the top of the list and having to pull far enough to make it happen. A normal scroll back to the top of it triggers it through inertia even if you've taken your finger off the screen.
- When refreshing I saw some strange cell rearranging going on... see video
- Is it possible to use the built-in pull to refresh here?

https://github.com/woocommerce/woocommerce-ios/assets/2472348/db1e106c-735b-4e0c-9548-2ac0b912e6b3


## Edit tax rate in Admin notice

- The notice is present and opens a web view ✅
- On a self hosted site I need to log in (I can't remember whether this is avoidable or not.)
- Using the iOS password manager integration caused the webview to be dismissed on iOS 15 (but not iOS 16.) That also happened when I tried to log in by typing the username and password, so I was unable to use the link on iOS 15.
- Perhaps we should link to the `Standard rates` tab, not the root `Tax options`?

https://github.com/woocommerce/woocommerce-ios/assets/2472348/11aa451e-60bf-4472-bc60-ff7d4b959e86


## Design queries
You mention that the design isn't finalised – I can see that it does contain country/state identifiers alongside the tax name though:
![Tax selection screen](https://github.com/woocommerce/woocommerce-ios/assets/2472348/f2d69924-3404-484f-93c1-d4a508f4da13)

Is this planned for a future PR? One thing I found while using it is that it's difficult to tell which tax is which, especially where they're likely to have the same (customer-facing) name, e.g. GST, VAT, Sales tax, etc.


I'll review the code now...

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/Taxes/NewTaxRateSelectorView.swift:
```diff
@@ -37,65 +37,46 @@ struct NewTaxRateSelectorView: View {
+                        }) {
+                            ForEach(viewModel.taxRateViewModels, id: \.id) { viewModel in
+                                TaxRateRow(viewModel: viewModel)
+                                Divider()
```
Small stylistic thought: perhaps the divider should be part of the row? Or a viewModifier on it? Not a big deal.

toupper replied:
That's a good question, I see the divider more as part of the list than the row; similar to the `separatorStyle` property on UITableView.

joshheald replied:
That makes sense. It should probably be a modifier (or default behaviour) on the List then, I think that's how it works with the built in List.

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/Taxes/NewTaxRateSelectorView.swift:
```diff
@@ -37,65 +37,46 @@ struct NewTaxRateSelectorView: View {
-
+            }
+            .onAppear {
+                viewModel.onLoadTrigger.send()
```
`onLoad` and `onAppear` have different semantic implications to me when reading the code. I expect `onLoad` to be called once in the lifecycle, and `onAppear` to be callable multiple times. Perhaps this is out of date in SwiftUI; I can't really remember which this will be.

If it'll happen more than once, I'd suggest renaming the view model's property.

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/Taxes/NewTaxRateSelectorViewModel.swift:
```diff
@@ -1,24 +1,164 @@
+                let hasNextPage = results.count == pageSize
+                onCompletion?(.success(hasNextPage))
+
+                if self.taxRateViewModels.isEmpty {
+                    self.syncState = .empty
+                } else if results.isEmpty {
+                    // We had results previously, but we didn't have any on this page request. Transition to results to stop the syncing visuals.
+                    self.transitionToResultsUpdatedState()
+                }
```
Minor question: would it be better if this was done elsewhere, e.g. in the `onCompletion` handler? Reading it, I'm guessing that the completion handler or something else will have set the `taxRateViewModels = []` if that's appropriate from the results, but I can't tell that for sure while reading the code.

I think we're probably forced into doing it here by the design of `PaginationTrackerDelegate`, but just in case we're not, I thought it worth asking.

toupper replied:
That's a good point right there. I also tried to find another place to do this, but the only place is when we're listening to changes in the results controller. Since these cases don't bring any change in the results controller, we are not notified there.

The completion handler is part of the `PaginationTracker`, which doesn't handle the view model state.

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/Taxes/NewTaxRateSelectorViewModel.swift:
```diff
@@ -1,24 +1,164 @@
+
+    func configureFirstPageLoad() {
+        // Listens only to the first emitted event.
+        onLoadTrigger.first()
```
So this constrains the `onAppear` to act like `onLoad`. Perhaps you could move the `.first()` to the definition of the `onLoadTrigger` property instead? That might remove the mental overhead while reading it here...

The comment could then be moved to the `onAppear` block in the view, to explain why we're linking appear to load.

toupper replied:
That makes sense, done in cd70c4ddda9998ca87e2d9af1ae724bf814e876d

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/Taxes/NewTaxRateSelectorView.swift:
```diff
@@ -37,65 +37,46 @@ struct NewTaxRateSelectorView: View {
-                                    .frame(width: nil, alignment: .trailing)
+                switch viewModel.syncState {
+                    case .results:
+                        RefreshableInfiniteScrollList(isLoading: viewModel.shouldShowBottomActivityIndicator,
```
I can see this is pre-existing, and probably the cause of the various paging/refreshing issues I mentioned in my comment. Perhaps a separate PR to improve it is warranted? It's not been touched for a year and a half, so given the pace of SwiftUI development, it might be possible to do better now....

toupper replied:
I removed that component in 210b19437e649b0372b11146bce27d3094c0fb2c As you mentioned, it was causing several issues due to the refresh mechanism, which is not necessary for this task. Fixing those issues is not worth the effort compared to the benefits it might bring.

joshheald reviewed (approved):
Since it's feature flagged, I've approved with various questions/suggestions. None are essential, just my thoughts on reading it. Feel free to take or leave any of them 😄

toupper committed:
Use order to sort rates

toupper committed:
Remove refresh capability from the list as it brings several issues

toupper committed:
Add more tax rate information

toupper committed:
Rename to make more explicit than it will only load once.

toupper committed:
Add clarifying comment

toupper committed:
Fix tests and copiable

toupper commented:
Thanks for your thorough review @joshheald!

As mentioned in other comments, I removed the refreshable capability of the list as it's not strictly required (just a nice-to-have) and brings some issues.
Apart from that I added the missing elements to the design and moved the bottom indicator down so it doesn't clash with the row divider.

joshheald commented:
> Thanks for your thorough review @joshheald!
> 
> As mentioned in other comments, I removed the refreshable capability of the list as it's not strictly required (just a nice-to-have) and brings some issues. Apart from that I added the missing elements to the design and moved the bottom indicator down so it doesn't clash with the row divider.

Looks good to me, thanks! Only thing that seems to be remaining is the difficulty I was having logging in on the webview using iOS 15. LMK if you need any help tracking it down ☺️ – there's nothing obvious in the logs though.

toupper commented:
> Looks good to me, thanks! Only thing that seems to be remaining is the difficulty I was having logging in on the webview using iOS 15. LMK if you need any help tracking it down ☺️ – there's nothing obvious in the logs though.

True, I created an issue to track that

https://github.com/woocommerce/woocommerce-ios/issues/10637
